src/main/scala/beam/agentsim/infrastructure/ZonalParkingManager.scala

```diff
@@ -1,468 +1,259 @@
+    * @param searchTree nested map structure assisting search for parking within a TAZ and by parking type
+    * @param stalls collection of all parking alternatives
+    * @param tazQuadTree lookup of all TAZs in this simulation
+    * @param random random generator used to sample a location from the TAZ for this stall
+    * @return a stall from the found ParkingZone, or a ParkingStall.DefaultStall
+    */
+  def incrementalParkingZoneSearch(
+    searchStartRadius: Double,
+    searchMaxRadius: Double,
+    destinationUTM: Location,
+    valueOfTime: Double,
+    parkingDuration: Double,
+    parkingTypes: Seq[ParkingType],
+    chargingInquiryData: Option[ChargingInquiryData[String, String]],
+    searchTree: ParkingZoneSearch.ZoneSearch,
+    stalls: Array[ParkingZone],
+    tazQuadTree: QuadTree[TAZ],
+    distanceFunction: (Coord, Coord) => Double,
+    random: Random
+  ): (ParkingZone, ParkingStall) = {
+
+    @tailrec
+    def _search(thisInnerRadius: Double, thisOuterRadius: Double): Option[(ParkingZone, ParkingStall)] = {
+      if (thisInnerRadius > searchMaxRadius) None
+      else {
+        val tazDistance: Map[TAZ, Double] =
+          tazQuadTree
+            .getRing(destinationUTM.getX, destinationUTM.getY, thisInnerRadius, thisOuterRadius)
+            .asScala
+            .map { taz =>
+              (taz, GeoUtils.distFormula(taz.coord, destinationUTM))
+            }
+            .toMap
+        val tazList: List[TAZ] = tazDistance.keys.toList
+
+        ParkingZoneSearch.find(
+          destinationUTM,
+          valueOfTime,
+          parkingDuration,
+          chargingInquiryData,
+          tazList,
+          parkingTypes,
+          searchTree,
+          stalls,
+          distanceFunction,
+          random
+        ) match {
+          case Some(
+              ParkingRanking.RankingAccumulator(
+                bestTAZ,
+                bestParkingType,
+                bestParkingZone,
+                bestCoord,
+                bestRankingValue
+              )
+              ) =>
+            val stallPrice: Double =
+              bestParkingZone.pricingModel
+                .map { PricingModel.evaluateParkingTicket(_, parkingDuration.toInt) }
+                .getOrElse(DefaultParkingPrice)
+
+            // create a new stall instance. you win!
+            val newStall = ParkingStall(
+              bestTAZ.tazId,
+              bestParkingZone.parkingZoneId,
+              bestCoord,
+              stallPrice,
+              bestParkingZone.chargingPointType,
+              bestParkingZone.pricingModel,
+              bestParkingType
+            )
+
+            Some { (bestParkingZone, newStall) }
+          case None =>
+            _search(thisOuterRadius, thisOuterRadius * SearchFactor)
+        }
+      }
+    }
+
+    _search(0, searchStartRadius) match {
+      case Some(result) =>
+        result
+      case None =>
+        val newStall = ParkingStall.emergencyStall(random)
```

Comment:
Should emergency stall be used if nothing is found? Or let the agent pass in whether it is an emergency, so that they can try again later if it's not an emergency?

Reply:
this is in fact the behavior we want, but the "emergency" is with respect to the supply, not the demand. the notion is, give them an expensive and far-away stall if no stalls are found. this will push agents to seek other modes in future iterations. 

perhaps the name "EmergencyStall" is misleading. it's "emergency" as in the simulation ran out of options to offer, and so we urgently need to fill one in. but if you can think of another name, i'm open to it.

Reply:
this is in fact the behavior we want, but the "emergency" is with respect to the supply, not the demand. the notion is, give them an expensive and far-away stall if no stalls are found. this will push agents to seek other modes in future iterations. 

perhaps the name "EmergencyStall" is misleading. it's "emergency" as in the simulation ran out of options to offer, and so we urgently need to fill one in. but if you can think of another name, i'm open to it.

Reply:
backup stall?

Reply:
I figured based on the code. Note that far away if it's an emergency seems counterproductive - I can understand costly, but far away defeats the purpose of meeting supply. 

As to the name...naming sucks, right! Emergency is good, and I was just verifying. Another alternative would be LastResortStall maybe?

Reply:
LastResortStall +1

Reply:
naming is NP-Complete ;-)

I'll switch it to __LastResortStall__.